test-tools/src/main/java/org/terracotta/utilities/test/logging/ConsoleAppenderCapture.java

```diff
@@ -0,0 +1,235 @@
+   * in the current Logback configuration having a {@link ConsoleAppender}.  Only events logged by threads
+   * for which the {@link MDC} contains a key generated by this constructor are captured.  If multiple
+   * {@code ConsoleAppender} instances are used (uncommon), the captures from all instances having the
+   * same target are aggregated.
+   * <p>
+   * If the current logging configuration does not append to {@code ConsoleAppender}, no capture is
+   * performed and {@link #getLogs()} will return an empty map.
+   * <p>
+   * If a {@code Logger}, through inheritance, appends to multiple instances of {@code ConsoleAppender},
+   * then multiple event captures may be observed.
+   * <p>
+   * The {@code ConsoleAppenderCapture} instance <b>must</b> be closed when no longer needed -- use
+   * try-with-resources.
+   *
+   * @param testId used as the value for the generated {@code MDC} key; if {@code null}, a value of
+   *               {@code present} is used
+   */
+  public ConsoleAppenderCapture(String testId) {
+    this.testId = (testId == null ? "present" : testId);
+
+    MDC.put(mdcKey, this.testId);
+    Runnable removeAppenders = () -> {
+    };
+    LoggerContext context = (LoggerContext)LoggerFactory.getILoggerFactory();
+
+    /*
+     * Scan for all loggers using the ConsoleAppender and add an appender to capture
+     * the output for the designated ConsoleAppender target.
+     */
+    for (ch.qos.logback.classic.Logger logger : context.getLoggerList()) {
+      Iterator<Appender<ILoggingEvent>> appenderIterator = logger.iteratorForAppenders();
+      while (appenderIterator.hasNext()) {
+        Appender<ILoggingEvent> iLoggingEventAppender = appenderIterator.next();
+        if (iLoggingEventAppender instanceof ConsoleAppender) {
+          String target = ((ConsoleAppender<?>)iLoggingEventAppender).getTarget();
+          ListAppender<ILoggingEvent> appender = appenderMap.computeIfAbsent(target, t -> {
+            ListAppender<ILoggingEvent> listAppender = new ListAppender<>();
+            listAppender.setContext(context);
+
+            listAppender.addFilter(new AbstractMatcherFilter<ILoggingEvent>() {
+              @Override
+              public FilterReply decide(ILoggingEvent iLoggingEvent) {
+                if (iLoggingEvent.getMDCPropertyMap().getOrDefault(mdcKey, "").equals(ConsoleAppenderCapture.this.testId)) {
+                  return FilterReply.ACCEPT;
+                } else {
+                  return FilterReply.DENY;      // Filter only applies to 'this' Appender -- can stop processing here
+                }
+              }
+            });
+
+            listAppender.start();
+            return listAppender;
+          });
+          logger.addAppender(appender);
+          Runnable r = removeAppenders;
+          removeAppenders = () -> {
+            r.run();
+            logger.detachAppender(appender);
+            appender.stop();
+          };
```

Comment:
what happens when 2 consecutive calls of factory methods with the same testId, whether in the same thread or not ? Should we only add the appender only once since putIfAbsent would return the existing instance ?

Reply:
Again, `ConsoleAppenderCapture` instances are independent regardless of the `testId` assigned.  `appenderMap` is an instance field with new, filtered `Appender` instances for the `ConsoleApenderCapture` instance.  (Each `ConsoleAppenderCapture` instance, at the moment, will have at most two (2) `ListAppender` instances -- one for `System.out` and/or one for `System.err`.)  If another `ConsoleAppenderCapture` instance is created over the same execution scope, it will add its own set of `ListAppender` instances adding its own MDC key for filtering purposes.  When a logging event is presented, every `Appender` instance chained from the `Logger` is called in the order in which it was added to the `Logger` -- in our case, the `ConsoleAppender` is called then the `ListAppender` instances added by each instance of `ConsoleAppenderCapture`.